sys/include/event.h

```diff
@@ -0,0 +1,214 @@
+ * @ingroup     sys
+ * @brief       Provides an Event loop
+ *
+ * This module offers an event queue framework not unlike libevent or libuev.
+ *
+ * An event queue is basically a FIFO queue of events, with some functions to
+ * efficiently and safely handle adding and getting events to / from such a
+ * queue.
+ * An event queue is bound to a thread, but any thread or ISR can put events
+ * into a queue.
+ *
+ * An event is a structure containing a pointer to an event handler. It can be
+ * extended to provide context or arguments to the handler.  It can also be
+ * embedded into existing structures (see examples).
+ *
+ * Compared to msg or mbox, this some fundamental differences:
+ *
+ * 1. events are "sender allocated". Unlike msg_send(), event_post() never
+ *    blocks or fails.
+ * 2. events contain everything necessary to handle them, thus a thread
+ *    processing the events of an event queue doesn't need to be changed in
+ *    order to support new event types.
+ * 3. events can be safely used (and actually perform best) when used within
+ *    one thread, e.g., in order to create a state-machine like process flow.
+ *    This is not (easily) possible using msg queues, as they might fill up.
+ * 4. an event can only be queued in one event queue at the same time.
+ *    Notifying many queues using only one event object is impossible.
+ *
+ * At the core, event_wait() uses thread flags to implement waiting for events
+ * to be queued. Thus event queues can be used safely and efficiently in combination
+ * with thread flags and msg queues.
+ *
+ * Examples:
+ *
+ *     // simple event handler
+ *     static void handler(event_t *event)
+ *     {
+ *        printf("triggered 0x%08x\n", (unsigned)event);
+ *     }
+ *
+ *     static event_t event = { .handler=handler };
+ *     static event_queue_t queue;
+ *
+ *     int main(void)
+ *     {
+ *         event_queue_init(&queue);
+ *         event_loop(&queue);
+ *     }
+ *
+ *     [...] event_post(&queue, &event);
+ *
+ *     // example for event extended event struct
+ *     typedef struct {
+ *         event_t super;
+ *         const char *text;
+ *     } custom_event_t;
+ *
+ *     static void custom_handler(event_t *event)
+ *     {
+ *         custom_event_t *custom_event = (custom_event_t *)event;
+ *         printf("triggered custom event with text: \"%s\"\n", custom_event->text);
+ *     }
+ *         static custom_event_t custom_event = { .super.callback=custom_handler, .text="CUSTOM EVENT" };
+ *
+ *     [...] event_post(&queue, &custom_event)
+ *
+ *
+ *
+ * @{
+ *
+ * @file
+ * @brief       Event API
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef EVENT_H
+#define EVENT_H
+
+#include <stdint.h>
+
+#include "irq.h"
+#include "thread_flags.h"
+#include "clist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef THREAD_FLAG_EVENT
+/**
+ * @brief   Thread flag use to notify available events in an event queue
+ */
+#define THREAD_FLAG_EVENT   0x1
```

Comment:
Params? Also, not sure how flag colisions are handled in `thread_flags` in general.

Reply:
I thought about it... making this configurable per queue adds 33% to the queue structure's size...
I hoped that we can get by with one queue per thread, in which case a globally configurable constant might be ok...

OK to leave as is and gather experience?

Reply:
> Params?

That was a typo... what I meant was "parans" ;-). As such I think you misunderstood the second half: I'm actually not sure if other modules using `thread_flags` won't collide with the flag you define here, so can you put my mind here at ease?

Reply:
> parans

fixed


> I'm actually not sure if other modules using thread_flags won't collide with the flag you define here, so can you put my mind here at ease?

Maybe half. :) Users of thread_flags might actually collide by using the same thread flag. I don't see that as a huge problem (unlike crashing msg ids), as thread flags are supposed to be used in a "notifying" way, meaning "there might be something you might want to look at".

As said, I'd like to gather some experience before deciding whether the flag number needs to be changed and/or made configurable.